src/ape/managers/chain.py

```diff
@@ -0,0 +1,360 @@
+
+
+class AccountHistory:
+    """
+    A container mapping account addresses to the transaction from the active session.
+    """
+
+    _map: Dict[AddressType, List[ReceiptAPI]] = {}
+
+    def __getitem__(self, address: Union[AddressAPI, AddressType, str]) -> List[ReceiptAPI]:
+        """
+        Get the list of transactions from the active session for the given address.
+
+        Args:
+            address (:class:`~ape.types.AddressType`): The sender of the desired transactions.
+
+        Returns:
+            List[:class:`~ape.api.providers.TransactionAPI`]: The list of transactions. If there
+            are no recorded transactions, returns the empty list.
+        """
+        address_key: AddressType = _convert(address, AddressType)
+        return self._map.get(address_key, [])
+
+    def append(self, txn_receipt: ReceiptAPI):
+        """
+        Add a transaction to the stored list for the given account address.
+
+        Raises:
+            :class:`~ape.exceptions.ChainError`: When trying to append a transaction
+              that is already in the list.
+
+        Args:
+            txn_receipt (:class:`~ape.api.providers.ReceiptAPI`): The transaction receipt to append.
+              **NOTE**: The receipt is accessible in the list returned from container[sender].
+        """
+        address = _convert(txn_receipt.sender, AddressType)
+        if address not in self._map:
+            self._map[address] = [txn_receipt]
+            return
+
+        if txn_receipt.txn_hash in [r.txn_hash for r in self._map[address]]:
+            raise ChainError(f"Transaction '{txn_receipt.txn_hash}' already known.")
+
+        self._map[address].append(txn_receipt)
+
+
+@dataclass
+class ChainManager:
+    """
+    A class for managing the state of the active blockchain.
+    Also handy for querying data about the chain and managing local caches.
+    Access the chain manager singleton from the root ``ape`` namespace.
+
+    Usage example::
+
+        from ape import chain
+    """
+
+    _networks: NetworkManager
+    _snapshots: List[SnapshotID] = []
+    _time_offset: int = 0
+    _chain_id_map: Dict[str, int] = {}
+
+    account_history: AccountHistory = AccountHistory()
+    """A mapping of transactions from the active session to the account responsible."""
+
+    def __repr__(self) -> str:
+        try:
+            return f"<ChainManager (chain_id={self.chain_id})>"
+        except ProviderNotConnectedError:
+            return "<ChainManager (disconnected)>"
+
+    @property
+    def provider(self) -> ProviderAPI:
+        """
+        The active :class:`~ape.api.providers.ProviderAPI`.
+
+        Raises:
+            :class:`~ape.exceptions.ProviderNotConnectedError`: When not connected
+              to a provider.
+        """
+
+        provider = self._networks.active_provider
+        if not provider:
+            raise ProviderNotConnectedError()
+
+        return provider
+
+    @property
```

Comment:
`cached_property`? so as not to continue initializing this class every time it is called

Reply:
Or just set it inside of `__post_init__`

Reply:
It is cached by network right now.
So if we have already made the request, the ID is already present in `self._chain_id_map`. Otherwise, we make the provider call. It is like `@cached_proeprty` but we are controlling more how we cache